app/team.go

```diff
@@ -769,17 +769,22 @@ func (a *App) AddTeamMember(teamId, userId string) (*model.TeamMember, *model.Ap
 	return teamMember, nil
 }
 
-func (a *App) AddTeamMembers(teamId string, userIds []string, userRequestorId string) ([]*model.TeamMember, *model.AppError) {
+func (a *App) AddTeamMembers(teamId string, userIds []string, userRequestorId string, graceful bool) ([]*model.TeamMember, []*model.AppError) {
```

Comment:
More discussion on golang's multierror https://github.com/pkg/errors/issues/15#issuecomment-355372589

Reply:
@ali-farooq0 
a. we need to return the errors, because they are displayed in the UI
b. I don't mind renaming graceful to besteffort ( I didn't like my name for this anyway :) )
c. All this weirdness is due to the need to keep backwards compatibility in the API
d. I am open to suggestions to improve this. I understand the rationale of your comment, just not sure how to make this prettier

Reply:
Ahh i see. So we need to inform the UI that SOME failed? do we need to inform them which failed? If not, then we can just introduce a new bool return value that indicates all members were successfully added. name could be like `addedAll` or `success` or something, and then return a new error from the API based on that bool. 

`func (a *App) AddTeamMembers(teamId string, userIds []string, userRequestorId string, graceful bool) (result []*model.TeamMember, success bool, appErr *model.AppError)`

OR

`func (a *App) AddTeamMembers(teamId string, userIds []string, userRequestorId string, graceful bool) (result []*model.TeamMember, failed []userIds, appErr *model.AppError)`

What do you think?

Reply:
we do something similar for our `KVCompareAndSet` api
https://github.com/mattermost/mattermost-server/blob/d0d6ce0a7014631c94cdcb3c01da233151799e0f/app/plugin_api.go#L708

Reply:
we not only need to notify WHICH for which users it failed, we also notify WHY it failed.
so returning just the userIDs is not enough :(

Reply:
no i mean, the json api detects that there were failed users and return `api.team.add_team_members_graceful.app_error`

Reply:
ah ok. where would you return the error strings? inside the params of AppError? That's what I've done initially, but changed it due to @jespino request

Reply:
Spoke to @reflog offline, suggested to return `([]*model.TeamMemberWithError, *model.AppError)`  that contains the error and/or teammember.

Reply:
I would even recommend sending that struct all the way to webapp, keep them consistent between server and webapp, what do you think?